src/Microsoft.Azure.WebJobs.Extensions.Redis/RedisExtensionConfigProvider.cs

```diff
@@ -62,14 +63,14 @@ public void Initialize(ExtensionConfigContext context)
 
         internal static IConnectionMultiplexer GetOrCreateConnectionMultiplexer(IConfiguration configuration, string connectionStringSetting, string clientName = "")
         {
-            string connectionString = RedisUtilities.ResolveConnectionString(configuration, connectionStringSetting);
-            return connectionMultiplexerCache.GetOrAdd(connectionString, (string cs) => CreateConnectionMultiplexer(cs, clientName));
+            return connectionMultiplexerCache.GetOrAdd(connectionStringSetting, (string css) => CreateConnectionMultiplexer(configuration, css, clientName));
         }
 
-        internal static IConnectionMultiplexer CreateConnectionMultiplexer(string connectionString, string clientName)
+        internal static IConnectionMultiplexer CreateConnectionMultiplexer(IConfiguration configuration, string connectionStringSetting, string clientName)
         {
-            ConfigurationOptions options = ConfigurationOptions.Parse(connectionString);
-            options.ClientName = string.Format(RedisUtilities.RedisClientNameFormat, clientName);
+            Task<ConfigurationOptions> optionsTask = RedisUtilities.ResolveConfigurationOptionsAsync(configuration, connectionStringSetting, clientName);
+            optionsTask.Wait();
```

Comment:
Calling `.Wait()` is a dangerous sync-over-async anti-pattern. Can we convert the `CreateConnectionMultiplexer()` method to async and use `await` instead?

Reply:
I was thinking about this as well but couldn't think of a great solution. Without this `.Wait()`, I will need to refactor some parts, but considering it is indeed a dangerous pattern, I'll make the changes to accommodate it.

At the moment, the  `RedisUtilities.ResolveConfigurationOptionsAsync` doesn't actually call an `async` method, but in the next PR #137 to add Entra ID support, the `ConfigureForAzureWithSystemAssignedManagedIdentityAsync` or the other managed identity configuration methods are all `async`.
 
In order to ensure that all connections from a function app to a single Redis instance use the same `IConnectionMultiplexer`, I created a `ConcurrentDictionary<IConnectionMultiplexer>`, and then use the `GetOrAdd` method with the lambda that creates the connection multiplexer if there isn't one. The problem is that the lambda does not allow for an `async` function, meaning that I can't use the `ConcurrentDictionary` to keep track of the connections across all threads.

It is possible to get around this by doing the following:
``` c#
internal static async Task<IConnectionMultiplexer> GetOrCreateConnectionMultiplexerAsync(IConfiguration configuration, string connectionStringSetting, string clientName = "")
{
    if (connectionMultiplexerCache.ContainsKey(connectionStringSetting))
    {
        connectionMultiplexerCache.TryGetValue(connectionStringSetting, out IConnectionMultiplexer connectionMultiplexer);
        return connectionMultiplexer;
    }
    else
    {
        IConnectionMultiplexer connectionMultiplexer = await CreateConnectionMultiplexerAsync(configuration, connectionStringSetting, clientName);
        return connectionMultiplexerCache.GetOrAdd(connectionStringSetting, connectionMultiplexer);
    }
}
```

The problem then comes with [creating the `RedisAsyncCollector`](https://github.com/Azure/azure-functions-redis-extension/blob/a6219188e8be801b43963e16581cb70b61ffefd8/src/Microsoft.Azure.WebJobs.Extensions.Redis/RedisExtensionConfigProvider.cs#L59), as it also requires a lambda to create and can't be an `async` function. This I've created a separate PR #139 to allow `async` creation of the `RedisAsyncCollector`.

The last holdout is [RedisScalerProvider](https://github.com/Azure/azure-functions-redis-extension/blob/a6219188e8be801b43963e16581cb70b61ffefd8/src/Microsoft.Azure.WebJobs.Extensions.Redis/RedisScalerProvider.cs#L31), which currently creates an `IConnectionMultiplexer` in its constructor. Since the constructor can't be `async`, I'll create a separate PR that refactors that logic to allow for calling `GetOrCreateConnectionMultiplexerAsync` during operation.

Reply:
Thanks to #139, #140, (and #142), updated this to be `async` all the way down